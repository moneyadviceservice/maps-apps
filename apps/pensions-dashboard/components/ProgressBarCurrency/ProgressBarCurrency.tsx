import { Paragraph } from '@maps-digital/shared/ui';

import { formatCurrency } from '@maps-react/pension-tools/utils/formatCurrency';

type Props = {
  amount: number;
  total: number;
  suffix?: string;
  testId?: string;
};

export const ProgressBarCurrency = ({
  amount,
  total,
  suffix,
  testId = 'progress-bar-currency',
}: Props) => {
  const value = Math.max(0, Math.min(amount, total));
  const progress = `${(value / total) * 100}%`;

  const text = suffix
    ? `${formatCurrency(value)} ${suffix}`
    : formatCurrency(value);

  return (
    <>
      <Paragraph
        data-testid={testId}
        className="text-base font-bold text-blue-700 md:mb-2"
      >
        {text}
      </Paragraph>
      <div
        aria-hidden="true"
        className="overflow-hidden h-9 md:h-12 bg-blue-700/20"
      >
        <div className="h-full bg-blue-700" style={{ width: progress }} />
      </div>
    </>
  );
};
