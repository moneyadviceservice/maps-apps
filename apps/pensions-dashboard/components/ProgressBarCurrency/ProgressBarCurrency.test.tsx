import { render, screen } from '@testing-library/react';

import { ProgressBarCurrency } from './ProgressBarCurrency';

import '@testing-library/jest-dom/extend-expect';

describe('ProgressBarCurrency', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ProgressBarCurrency amount={123.45} total={1000} suffix="a month" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('handles zero amount', () => {
    render(<ProgressBarCurrency amount={0} total={1000} suffix="a month" />);
    expect(screen.getByTestId('progress-bar-currency')).toHaveTextContent(
      '£0.00 a month',
    );
  });

  it('handles whole number amounts', () => {
    render(<ProgressBarCurrency amount={100} total={1000} suffix="a month" />);
    expect(screen.getByTestId('progress-bar-currency')).toHaveTextContent(
      '£100.00 a month',
    );
  });

  it('rounds long decimal amount to currency format', () => {
    render(
      <ProgressBarCurrency amount={123.4567} total={1000} suffix="a month" />,
    );
    expect(screen.getByTestId('progress-bar-currency')).toHaveTextContent(
      '£123.46 a month',
    );
  });

  it('handles amounts greater than the total', () => {
    render(<ProgressBarCurrency amount={2000} total={1000} suffix="a month" />);

    expect(screen.getByTestId('progress-bar-currency')).toHaveTextContent(
      `£1,000.00 a month`,
    );
  });

  it('handles negative amounts', () => {
    render(
      <ProgressBarCurrency amount={-123.45} total={1000} suffix="a month" />,
    );
    expect(screen.getByTestId('progress-bar-currency')).toHaveTextContent(
      '£0.00 a month',
    );
  });

  it('handles when no suffix is provided', () => {
    render(<ProgressBarCurrency amount={123.45} total={1000} />);
    expect(screen.getByTestId('progress-bar-currency')).toHaveTextContent(
      '£123.45',
    );
  });
});
