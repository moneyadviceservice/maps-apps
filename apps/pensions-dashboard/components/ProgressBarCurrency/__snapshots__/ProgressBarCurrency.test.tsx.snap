// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBarCurrency renders correctly 1`] = `
<div>
  <p
    class="mb-4 text-base font-bold text-blue-700 md:mb-2"
    data-testid="progress-bar-currency"
  >
    Â£123.45 a month
  </p>
  <div
    aria-hidden="true"
    class="overflow-hidden h-9 md:h-12 bg-blue-700/20"
  >
    <div
      class="h-full bg-blue-700"
      style="width: 12.345%;"
    />
  </div>
</div>
`;
